extension/src/router_channel.rs

```diff
@@ -450,9 +451,12 @@ impl RouterChannel {
       header_buffer.extend(b"\r\n");
       tx.send(Ok(Frame::data(header_buffer.into()))).await?;
 
-      let mut encoder: Option<GzEncoder<Vec<u8>>> = None;
+      let mut encoder: Option<GzEncoder<Writer<BytesMut>>> = None;
       if app_res_will_compress {
-        encoder = Some(GzEncoder::new(Vec::new(), flate2::Compression::default()));
+        encoder = Some(GzEncoder::new(
+          BytesMut::new().writer(),
```

Comment:
Perhaps there is a method to guess the size of the compressed data and make use of `BytesMut::with_capacity`?

Reply:
If we have to make a guess to reserve space and prevent most reallocs but also allow a realloc if it needs more space then we can just guess 15-50% of the content-length when available. If content-length is not set then we can start with a guess of, say, 10-20 KB to avoid overly small reallocs.

Reply:
But actually, we are supposed to be streaming this so the size in some cases will be based on the chunk size so we should only ever need, say, 10-20 KB and could add a loop to turn large chunks into small chunks of compressed output.

Reply:
Maybe there are other good reasons to do it, but I'm not sure a loop to shrink chunk size would matter in this case. Due to the fact that we are encoding the chunk data, we are forced to allocate at least as much (in sum) as is needed for the total compressed data over the course of the entire stream. The difference between this implementation and the previous implementation is the `Bytes` buffer does not get "reused" in the encoder before being copied into `Frame::data`. With each chunk, bytes are written to the buffer once and then immediately split off and moved into `Frame::data`, avoiding the need to copy the `Vec<u8>` and convert back to `Bytes` (which also usually requires a heap allocation).

To me the question is, if we knew the total size of the compressed data, would we want to allocated it in one shot upfront and then not have to allocate again. Or would we rather amortize? We will probably have to do some benchmarking on that trade-off.

Reply:
If we get set up to do some benchmarking here, one option we could explore is keeping track of how much we have filled a bytes buffer and then allocating and swapping in a new one when the time is right. Something like this:
```rust
let bytes = std::mem::replace(
    encoder.get_mut(),
    BytesMut::with_capacity(optimal_size).writer(),
).into_inner().into();
```